src/library_dylink.js

```diff
@@ -256,60 +256,106 @@ var LibraryDylink = {
   // returns the side module metadata as an object
   // { memorySize, memoryAlign, tableSize, tableAlign, neededDynlibs}
   $getDylinkMetadata: function(binary) {
-    var next = 0;
+    var offset = 0;
+    var end = 0;
+
+    function getU8() {
+      return binary[offset++];
+    }
+
     function getLEB() {
       var ret = 0;
       var mul = 1;
       while (1) {
-        var byte = binary[next++];
+        var byte = binary[offset++];
         ret += ((byte & 0x7f) * mul);
         mul *= 0x80;
         if (!(byte & 0x80)) break;
       }
       return ret;
     }
 
+    function getString() {
+      var len = getLEB();
+      offset += len;
+      return UTF8ArrayToString(binary, offset - len, len);
+    }
+
+    var name = 'dylink.0';
     if (binary instanceof WebAssembly.Module) {
-      var dylinkSection = WebAssembly.Module.customSections(binary, "dylink");
+      var dylinkSection = WebAssembly.Module.customSections(binary, name);
+      if (dylinkSection.length === 0) {
+        name = 'dylink'
+        dylinkSection = WebAssembly.Module.customSections(binary, name);
+      }
       assert(dylinkSection.length != 0, 'need dylink section');
       binary = new Uint8Array(dylinkSection[0]);
+      end = binary.length
     } else {
       var int32View = new Uint32Array(new Uint8Array(binary.subarray(0, 24)).buffer);
       assert(int32View[0] == 0x6d736100, 'need to see wasm magic number'); // \0asm
-      // we should see the dylink section right after the magic number and wasm version
-      assert(binary[8] === 0, 'need the dylink section to be first')
-      next = 9;
-      getLEB(); //section size
-      assert(binary[next] === 6);                 next++; // size of "dylink" string
-      assert(binary[next] === 'd'.charCodeAt(0)); next++;
-      assert(binary[next] === 'y'.charCodeAt(0)); next++;
-      assert(binary[next] === 'l'.charCodeAt(0)); next++;
-      assert(binary[next] === 'i'.charCodeAt(0)); next++;
-      assert(binary[next] === 'n'.charCodeAt(0)); next++;
-      assert(binary[next] === 'k'.charCodeAt(0)); next++;
-    }
-
-    var customSection = {};
-    customSection.memorySize = getLEB();
-    customSection.memoryAlign = getLEB();
-    customSection.tableSize = getLEB();
-    customSection.tableAlign = getLEB();
+      // we should see the dylink custom section right after the magic number and wasm version
+      if (binary[8] !== 0) {
+        throw new Error('invalid shared library');
```

Comment:
This should probably be consistent with the other checks for valid dylibs. It looks like in this function we have `assert()`, `throw new Error`, and `err()`, all for different ways the library could be invalid.

Reply:
+1 for `assert`s instead of `throw`s.

Reply:
Does `assert` only work in ASSERTIONS builds? I think we need the validity checking even in release builds, because the error is based on externally-supplied invalid input, rather than an internal invariant.

Reply:
`assert` works in all builds. We optionally guard it behind `#ifdef ASSERTIONS` if they are optional. I agree that here they should be mandatory.

Reply:
Sounds good.  I will revert to using assert.  However the idea of assert (at least in native code) is that it should not do anything in release builds, but is used in debug build to check assumptions.   We might want to revisit this at some point.

Reply:
Yes, the term is maybe not optimal. Might be worth refactoring, like perhaps `assert() => check()` or such? (and leave `ASSERTIONS` as it is)